docs/root/configuration/http/http_filters/router_filter.rst

```diff
@@ -417,6 +417,14 @@ owning HTTP connection manager.
   rq_direct_response, Counter, Total requests that resulted in a direct response
   rq_total, Counter, Total routed requests
   rq_reset_after_downstream_response_started, Counter, Total requests that were reset after downstream response had started
+  passthrough_internal_redirect_stream_request, Counter, Total unsuccessful internal redirects due to the request stream not being complete
```

Comment:
one high level question before I dive in, which is to tag @jmarantz and check if we're keen on adding per-route per-failure stats here.  I _think_ the state of things is this will add a bunch of per-route memory overhead and if there's a way to ameliorate that.

Reply:
Thanks for the tag. I am not in a position to judge whether a stat is worth the memory tax or not, at least not without reading the whole PR :)

My thing is just to call out that the costs are on the order of 100 bytes per stat, multiplied by # of stats and cardinality of the thing you are attaching the new stats to. That's assuming you are composing stat-names largely with otherwise-existing tokens, so the amortized cost of the symbolized stats is small compared to the stat structures. If you are using Dynamic stats then the cost of the names is much higher.

If we think that's worth it, great. If it might only be of interest in some situations, there is precedent for capturing the stats in an absl::optional block and controlling whether they are instantiated with a runtime setting.

Reply:
I'm a bit rusty on the codebase by now, but I was wondering about lazily creating these metrics only if an internal redirect policy was configured in the first place? (also rusty on my C++ but) I imagine that looking like a unique pointer to InternalRedirectPolicy counters, so that it no policy is configured it's just `nullptr`. Not exactly "pay for what you use" since it still carries the cost of a pointer, but much smaller. It would also bring the cost of pointer lookup to increment the stats, but I think the design of InternalRedirectPolicy is more of handling exceptional behavior than really being the hot path?

(Have also thought of proposing to move the internal redirect stuff to a dedicated HTTP filter, then it'd really be pay-for-what-you-use if unconfigured and/or compiled out. So far I think it's feasible since it just needs filter state, request buffering, and stream recreate but idk the unknowns.)

Reply:
lazy-initializing stats sgtm.

Reply:
+1 that'd be awesome if you're up for it.  given there already are some redirect stats and the delta isn't large I could also be ok as-is - lmk